File: mytest/test_pg_xccl_send_recv.py
"""
Demonstrate ProcessGroupXCCL::send/recv via torch.distributed.

C++ equivalent:
    auto pg = c10::make_intrusive<ProcessGroupXCCL>(store, rank, size);
    std::vector<at::Tensor> tensors = {tensor};
    auto work = pg->send(tensors, dst, tag);   // or pg->recv(tensors, src, tag)
    work->wait();
"""
import os
import torch
import torch.distributed as dist
import torch.multiprocessing as mp

def worker(rank, world_size, file_name):
    # C++: auto store = c10::make_intrusive<FileStore>(path, world_size);
    store = dist.FileStore(file_name, world_size)

    # C++: auto pg = c10::make_intrusive<ProcessGroupXCCL>(store, rank, size);
    dist.init_process_group(backend="xccl", rank=rank, world_size=world_size, store=store)
    pg = dist.distributed_c10d._get_default_group()
    device = f"xpu:{rank}"

    if rank == 0:
        # C++: pg->send({tensor}, dstRank=1, tag=0)->wait();
        tensor = torch.full((10,), 42.0, device=device)
        dist.send(tensor, dst=1)
        print(f"[rank {rank}] sent: {tensor}")
    else:
        # C++: pg->recv({tensor}, srcRank=0, tag=0)->wait();
        tensor = torch.empty((10,), device=device)
        dist.recv(tensor, src=0)

        expected = torch.full((10,), 42.0, device=device)
        if tensor.equal(expected):
            print(f"[rank {rank}] PASS: received tensor matches")
        else:
            print(f"[rank {rank}] FAIL: got {tensor}, expected {expected}")

    # C++: dist->destroy_process_group();
    dist.destroy_process_group()

if __name__ == "__main__":
    world_size = min(torch.xpu.device_count(), 2)
    file_name = "/tmp/test_pg_xccl_send_recv"
    mp.spawn(worker, args=(world_size, file_name), nprocs=world_size)
    try:
        os.remove(file_name)
    except OSError:
        pass

File: mytest/test_pg_xccl_send_recv.cpp
// Demonstrate ProcessGroupXCCL::send/recv from C++.
//
// Build:
//   icpx -fsycl -DUSE_C10D_XCCL -Doverride= \
//     -I/path/to/torch-xpu-ops/src \
//     -I/path/to/torch/include \
//     -I/path/to/torch/include/torch/csrc/api/include \
//     -D_GLIBCXX_USE_CXX11_ABI=1 \
//     test_pg_xccl_send_recv.cpp \
//     -L/path/to/torch/lib \
//     -ltorch_xpu -ltorch_cpu -ltorch_python -ltorch -lc10_xpu -lc10 \
//     -o test_pg_xccl_send_recv
//
// Run:
//   export LD_LIBRARY_PATH=/path/to/torch/lib
//   mpirun -np 2 ./test_pg_xccl_send_recv <rank> <world_size>
//   # or manually in two terminals:
//   ./test_pg_xccl_send_recv 0 2   # terminal 1
//   ./test_pg_xccl_send_recv 1 2   # terminal 2

#include <ATen/ATen.h>
#include <c10/core/Device.h>
#include <c10/util/intrusive_ptr.h>
#include <torch/csrc/distributed/c10d/FileStore.hpp>
#include <xccl/ProcessGroupXCCL.hpp>
#include <c10/xpu/XPUStream.h>
#include <c10/xpu/impl/XPUGuardImpl.h>

#include <iostream>
#include <memory>
#include <string>

using namespace c10d;

int main(int argc, char* argv[]) {
  if (argc < 3) {
    std::cerr << "Usage: " << argv[0] << " <rank> <world_size>\n";
    return 1;
  }
  int rank = std::stoi(argv[1]);
  int world_size = std::stoi(argv[2]);

  // === 1. Create Store and ProcessGroupXCCL ===
  auto store = c10::make_intrusive<FileStore>("/tmp/pg_demo", world_size);

  auto pg = c10::make_intrusive<ProcessGroupXCCL>(store, rank, world_size);

  // === 2. Create XPU tensor ===
  auto device = torch::Device(torch::kXPU, rank);
  torch::Tensor tensor;
  if (rank == 0) {
    tensor = torch::full({10, 5}, 42.0f,
                         torch::TensorOptions().device(device));
    std::cout << "[rank " << rank << "] sending tensor of shape "
              << tensor.sizes() << std::endl;
  } else {
    tensor = torch::empty({10, 5}, torch::TensorOptions().device(device));
  }

  // === 3. send / recv via ProcessGroupXCCL ===
  //     pg->send({tensor}, dstRank, tag)->wait()
  //     pg->recv({tensor}, srcRank, tag)->wait()
  // send/recv are protected in ProcessGroupXCCL but public in Backend.
  // Cast to Backend* to call through the base class interface.
  auto* backend = static_cast<c10d::Backend*>(pg.get());

  std::vector<at::Tensor> tensors = {tensor};
  c10::intrusive_ptr<Work> work;

  if (rank == 0) {
    work = backend->send(tensors, /*dstRank=*/1, /*tag=*/0);
  } else {
    work = backend->recv(tensors, /*srcRank=*/0, /*tag=*/0);
  }
  work->wait();
  c10::xpu::getCurrentXPUStream(device.index()).queue().wait();

  // === 4. Verify ===
  if (rank == 1) {
    auto expected = torch::full({10, 5}, 42.0f,
                                torch::TensorOptions().device(device));
    if (tensor.equal(expected)) {
      std::cout << "[rank " << rank << "] PASS: received tensor matches\n";
    } else {
      std::cout << "[rank " << rank << "] FAIL: data mismatch\n";
      return 1;
    }
  } else {
    std::cout << "[rank " << rank << "] PASS: send completed\n";
  }

  return 0;
}
